File: benchmarks/fannkuch/fannkuch.go
package main

import "fmt"

func fannkuch(n int) int {
	perm := make([]int, n)
	perm1 := make([]int, n)
	count := make([]int, n)

	for i := 0; i < n; i++ {
		perm1[i] = i
	}

	maxFlips := 0
	checksum := 0
	sign := 1
	r := n

	for {
		// Generate next permutation
		for r != 1 {
			count[r-1] = r
			r--
		}

		// Copy perm1 to perm
		for i := 0; i < n; i++ {
			perm[i] = perm1[i]
		}

		// Count flips
		flips := 0
		for perm[0] != 0 {
			k := perm[0] + 1
			// Reverse first k elements
			for i, j := 0, k-1; i < j; i, j = i+1, j-1 {
				perm[i], perm[j] = perm[j], perm[i]
			}
			flips++
		}

		if flips > maxFlips {
			maxFlips = flips
		}
		checksum += sign * flips
		sign = -sign

		// Next permutation
		for {
			if r == n {
				fmt.Println(checksum)
				return maxFlips
			}
			// Rotate perm1[0..r]
			p0 := perm1[0]
			for i := 0; i < r; i++ {
				perm1[i] = perm1[i+1]
			}
			perm1[r] = p0

			count[r]--
			if count[r] > 0 {
				break
			}
			r++
		}
	}
}

func main() {
	n := 8
	result := fannkuch(n)
	fmt.Printf("Pfannkuchen(%d) = %d\n", n, result)
}

File: benchmarks/fannkuch/fannkuch.py
def fannkuch(n):
    perm = [0] * n
    perm1 = list(range(n))
    count = [0] * n

    max_flips = 0
    checksum = 0
    sign = 1
    r = n

    while True:
        # Generate next permutation
        while r != 1:
            count[r - 1] = r
            r -= 1

        # Copy perm1 to perm
        perm[:] = perm1[:]

        # Count flips
        flips = 0
        while perm[0] != 0:
            k = perm[0] + 1
            # Reverse first k elements
            perm[:k] = perm[:k][::-1]
            flips += 1

        if flips > max_flips:
            max_flips = flips
        checksum += sign * flips
        sign = -sign

        # Next permutation
        while True:
            if r == n:
                print(checksum)
                return max_flips
            # Rotate perm1[0..r]
            p0 = perm1[0]
            for i in range(r):
                perm1[i] = perm1[i + 1]
            perm1[r] = p0

            count[r] -= 1
            if count[r] > 0:
                break
            r += 1

if __name__ == "__main__":
    n = 10
    result = fannkuch(n)
    print(f"Pfannkuchen({n}) = {result}")
